fix host callback unit test

psychocoderHPC opened the pull request:
bug introduced with #1266

Unit tests are written so that the status of a queue is well-known and should not return false negatives.
This PR reverts the changes from #1266. If we observe still issues with HIP then something is wrong in the callback implementation.

bernhardmgruber commented on test/unit/queue/src/QueueTest.cpp:
```diff
@@ -65,7 +43,13 @@ TEMPLATE_LIST_TEST_CASE("queueCallbackIsWorking", "[queue]", TestQueues)
     alpaka::enqueue(f.m_queue, [&]() { promise.set_value(true); });
 
-    LOOPED_CHECK(30, 100, promise.get_future().get());
+    // A non-blocking queue will always stay empty because the task has been executed immediately.
```
I disagree. A non-blocking queue will be non-empty until the background thread has finished executing the task.

psychocoderHPC replied:
The background thread is an alpaka workaround which should not affect the behavior of `enqueue`. In a blocking queue, we need the guarantee that if no other thread is running the queue finished all tasks before.

bernhardmgruber replied:
The same goes for CUDA. You enqueue a host func into the CUDA stream. Is the CUDA stream empty inside the host func? IMO no.

bernhardmgruber commented on test/unit/queue/src/QueueTest.cpp:
```diff
@@ -65,7 +43,13 @@ TEMPLATE_LIST_TEST_CASE("queueCallbackIsWorking", "[queue]", TestQueues)
 
-    LOOPED_CHECK(30, 100, promise.get_future().get());
+    // A non-blocking queue will always stay empty because the task has been executed immediately.
+    if(!alpaka::test::IsBlockingQueue<typename Fixture::Queue>::value)
+    {
+        alpaka::wait(f.m_queue);
+    }
+
+    CHECK(promise.get_future().get());
```
No wait should be necessary, since the `std::future::get()` blocks until the promise is satisfied, and thus synchronizes with the task.

psychocoderHPC replied:
yes you are right

bernhardmgruber commented on test/unit/queue/src/QueueTest.cpp:
```diff
@@ -101,8 +85,11 @@ TEMPLATE_LIST_TEST_CASE(
+            CHECK(!alpaka::empty(f.m_queue));
             std::this_thread::sleep_for(std::chrono::milliseconds(100u));
+            // Check again, the queue must stay not empty.
+            CHECK(!alpaka::empty(f.m_queue));
```
So this is just a simplification of the test without a change in behavior?

psychocoderHPC replied:
No this is extending the test, we need to check that the queue stays non-empty as long as the task within the callback queue is executed.

bernhardmgruber replied:
The previous test did:
```
CHECK(!empty(queue));
wait 100ms
CHECK(!empty(queue));
wait 100ms
CHECK(!empty(queue));
wait 100ms
CHECK(!empty(queue));
wait 100ms
... // 30 times
```
The new test does only:
```
CHECK(!empty(queue));
wait 100ms
CHECK(!empty(queue));
```
IMO this is a simplification. What additional behavior is your new version covering?

psychocoderHPC replied:
The previous implementation was only repeating the command in the case where the first check returned that the queue is empty but we are in the host function which is blocking the queue therefore it must be `!empty` else something is wrong.
The second check validates CUDA/HIP or our bookkeeping with the callback thread that if a task which is blocking the execution on the host is really blocking the queue too.

bernhardmgruber replied:
Oh yes, you are completely right! Sorry, my mistake! So yes, your proposed change is better than what we had before!

bernhardmgruber commented on test/unit/queue/src/QueueTest.cpp:
```diff
@@ -113,7 +100,10 @@ TEMPLATE_LIST_TEST_CASE(
     CHECK(callbackFinished.load() == true);
-    LOOPED_CHECK(30, 100, alpaka::empty(f.m_queue));
+
+    // The blocking queue must be empty because we wait for the non-blocking queue and the blocking is synchronized
+    // implicit.
```
```suggestion
    // The blocking queue must be empty because we wait for the non-blocking queue and the blocking one is synchronized
    // implicitly.
```

bernhardmgruber reviewed (commented):
I don't see what this PR fixes. I could not spot the bug that was present before.

psychocoderHPC committed:
fix host callback unit test

bug introduced with #1266

Unit tests are written so that the status of a queue is well-known and should not return false negatives.
This PR reverts the changes from #1266. If we observe still issues with HIP then something is wrong in the callback implementation.

psychocoderHPC commented:
@bernhard PR is ready to be merged.
